external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -445,197 +529,68 @@ private[kafka010] case class InternalKafkaConsumer(
     }
   }
 
   /**
-   * Get a cached consumer for groupId, assigned to topic and partition.
+   * Get a data reader for groupId, assigned to topic and partition.
    * If matching consumer doesn't already exist, will be created using kafkaParams.
-   * The returned consumer must be released explicitly using [[KafkaDataConsumer.release()]].
-   *
-   * Note: This method guarantees that the consumer returned is not currently in use by any one
-   * else. Within this guarantee, this method will make a best effort attempt to re-use consumers by
-   * caching them and tracking when they are in use.
+   * The returned data reader must be released explicitly.
    */
   def acquire(
       topicPartition: TopicPartition,
-      kafkaParams: ju.Map[String, Object],
-      useCache: Boolean): KafkaDataConsumer = synchronized {
-    val key = new CacheKey(topicPartition, kafkaParams)
-    val existingInternalConsumer = cache.get(key)
-
-    lazy val newInternalConsumer = new InternalKafkaConsumer(topicPartition, kafkaParams)
-
+      kafkaParams: ju.Map[String, Object]): KafkaDataConsumer = {
     if (TaskContext.get != null && TaskContext.get.attemptNumber >= 1) {
-      // If this is reattempt at running the task, then invalidate cached consumer if any and
-      // start with a new one.
-      if (existingInternalConsumer != null) {
-        // Consumer exists in cache. If its in use, mark it for closing later, or close it now.
-        if (existingInternalConsumer.inUse) {
-          existingInternalConsumer.markedForClose = true
-        } else {
-          existingInternalConsumer.close()
-        }
-      }
-      cache.remove(key)  // Invalidate the cache in any case
-      NonCachedKafkaDataConsumer(newInternalConsumer)
-
-    } else if (!useCache) {
-      // If planner asks to not reuse consumers, then do not use it, return a new consumer
-      NonCachedKafkaDataConsumer(newInternalConsumer)
+      // If this is reattempt at running the task, then invalidate cached consumer if any.
 
-    } else if (existingInternalConsumer == null) {
-      // If consumer is not already cached, then put a new in the cache and return it
-      cache.put(key, newInternalConsumer)
-      newInternalConsumer.inUse = true
-      CachedKafkaDataConsumer(newInternalConsumer)
+      val cacheKey = new CacheKey(topicPartition, kafkaParams)
 
-    } else if (existingInternalConsumer.inUse) {
-      // If consumer is already cached but is currently in use, then return a new consumer
-      NonCachedKafkaDataConsumer(newInternalConsumer)
-
-    } else {
-      // If consumer is already cached and is currently not in use, then return that consumer
-      existingInternalConsumer.inUse = true
-      CachedKafkaDataConsumer(existingInternalConsumer)
+      // invalidate all fetched data for the key
+      // sadly we can't pinpoint specific data and invalidate cause we don't have unique id
+      fetchedDataPool.invalidate(cacheKey)
```

Comment:
Just to double check does this invalidate + create a new instance? This is essential for delegation token.

Reply:
I need to refresh the context of this PR but if my memory is right, it just invalidates consumers belong to the specific key, and borrowed consumers will not be closed at that time.

Btw I'm not sure I'm following your point. I haven't take a deep look with Kafka delegation token, but isn't it essential to deal with delegation token whenever creating consumer?

Reply:
When new delegation token is obtained it won't affect the cached consumers (same on producer side). A consumer with and old token will be used until it's not throwing an exception. When an exception is thrown the old consumer has to be invalidated (but not necessarily closed) and at the same time new consumer has to be created (which will pick up the latest token). Just a side note the producer side is not yet implemented like this because https://github.com/apache/spark/pull/23956 is depending on https://github.com/apache/spark/pull/19096 and not yet merged.

I know this solution is not optimal and because of that I'm planning to implement a feature to invalidate all consumers in case of new token arrival. I think this makes sense after this has been merged. But for now we have to ensure the original solution still works...

Reply:
Is there any specific exception when delegation token has expired? If then we can try to deal with invalidation & creation of consumer if it catches the exception. If it doesn't, we may be required to invalidate consumer once any exception is thrown.

Yes, assuming we have retry logic for every usage of consumer. Otherwise we should add it, or simply let just task fails and being retried by Spark side.

Reply:
So back to this. As a general comment I'm just adding end-to-end delegation token test with KDC in https://github.com/apache/spark/pull/25477. That could be a good sign if that one doesn't break (though that's not testing failure scenario).

I think invalidating + creating a new consumer is a must without knowing what kind of exception has been thrown. This feature already exists in the actual implementation and we have to keep this.

Just a side note I want to add similar feature on the producer side in https://github.com/apache/spark/pull/23956 but it's depending on another PR which is not moving :/

Reply:
This is same as current: we invalidate consumers per key according to any task failure, so I guess this is OK unless we would want to have better solution.

Even we can pass the fact whether the exception occurs or not when returning borrowed consumer, to let pool destroy the consumer instead of pooling it.
`KafkaDataConsumer.release()`.

Reply:
Had a deeper look and it should provide the same behavior just like before.
I think it's quite an important feature which is covered neither in DStreams nor in Structured Streaming. Let me check if I can add such test for both.

Reply:
I've filed https://github.com/apache/spark/pull/25582 to fill this gap and we can make sure this still works.

Reply:
OK let's follow up #25582 and see whether the tests pass with current master (we may want to verify this first) and also pass with current patch.

Reply:
Also reflected #25582 here as well.